### PR TITLE
[WIP] Preparing code for adding nanmean function

### PR DESCRIPTION
**Pull request by krshrimali**
Work in Progress: This PR is to prepare work to add `nanmean` function to PyTorch (ReduceOps).

**TODO:**

* [x] Initial (a.k.a manual) testing after successful builds.
* [ ] Add tests to test/test_torch.py after creating a draft PR to PyTorch.

**Questions:**

* Can we avoid using `at::sum` and instead do this in `combine` of `NaNMeanOps`?
* Should we use `at::_isnan` format or `at::isnan`? What's the difference?

### CODE REVIEW

**Comment by krshrimali** [comment]
This script does the task:

```python
import torch

y = torch.tensor([[1, float('nan'), 3], [float('nan'), 5, 6]])
without_nans = torch.where(torch.isnan(y), torch.zeros_like(y), y)
if (y.device.type == 'cpu'):
    dim_prod = torch.sum(~torch.isnan(y), axis=0, keepdims=False)
    result = torch.zeros_like(y)
    final = torch.sum(without_nans, axis=0, keepdims=False, out=result).true_divide(dim_prod)
    print(y, final)
```

Output:
```bash
# Input Tensor
tensor([[1., nan, 3.],
        [nan, 5., 6.]])
# Output (nanmean of input tensor) Tensor 
tensor([1.0000, 5.0000, 4.5000])
```

**Comment by krshrimali** [comment]
> This script does the task:
> 
> ```python
> import torch
> 
> y = torch.tensor([[1, float('nan'), 3], [float('nan'), 5, 6]])
> without_nans = torch.where(torch.isnan(y), torch.zeros_like(y), y)
> if (y.device.type == 'cpu'):
>     dim_prod = torch.sum(~torch.isnan(y), axis=0, keepdims=False)
>     result = torch.zeros_like(y)
>     final = torch.sum(without_nans, axis=0, keepdims=False, out=result).true_divide(dim_prod)
>     print(y, final)
> ```
> 
> Output:
> 
> ```shell
> # Input Tensor
> tensor([[1., nan, 3.],
>         [nan, 5., 6.]])
> # Output (nanmean of input tensor) Tensor 
> tensor([1.0000, 5.0000, 4.5000])
> ```

The target is to revolve around this to implement `nanmean`. I'll keep testing with possibilities of `keepdims` and `axis`.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -498,6 +498,50 @@ Tensor& mean_out(Tensor& result, const Tensor& self, DimnameList dim,
...
+  TORCH_CHECK(
+      at::isFloatingType(scalarType) || at::isComplexType(scalarType),
+      "Can only calculate the mean of floating types. Got ",
+      toString(scalarType),
+      " instead.");
+  ScalarType dtype = get_dtype(result, self, opt_dtype, true);
+  Tensor without_nans = torch.where(torch.isnan(self), torch.zeros_like(self), self);
+  if (self.device().is_cpu()) {
+    int64_t dim_prod = at::sum(~self.isnan(), axis=dim, keepdims=keepdim, dtype=dtype);
+    at::sum_out(result, without_nans, dim, keepdim, dtype).div_(dim_prod);
```
Comment (krshrimali) [inline aten/src/ATen/native/ReduceOps.cpp:512]: `.div_` throws a `RuntimeWarning` (and will soon through an error instead, see: https://github.com/pytorch/pytorch/pull/38620) in Python. Let's change this to `true_divide` after a little testing.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -498,6 +498,50 @@ Tensor& mean_out(Tensor& result, const Tensor& self, DimnameList dim,
...
+  }
+  /*
+  auto iter = make_reduction("nanmean", result, self, dim, keepdim, dtype);
+  if (iter.numel() == 0) {
+    result.fill_(std::numeric_limits<double>::quiet_NaN());
+  } else {
+    nanmean_stub(iter.device_type(), iter);
+  }
+  return result;;
+  */
```
Comment (krshrimali) [inline aten/src/ATen/native/ReduceOps.cpp:525]: TODO. Need to understand what are we doing here. My best guess:

```cpp
if (number of elements are 0) {
    // Fill result with NAN (if float type), else 0 - for other types, check: https://en.cppreference.com/w/cpp/types/numeric_limits/quiet_NaN
} else {
    // Call the CUDA kernel defined in aten/src/ATen/native/cuda/ReduceMomentKernel.cu
}
```

File: aten/src/ATen/native/cuda/ReduceMomentKernel.cu [hunk 3]
```diff
@@ -62,6 +62,12 @@ static void mean_kernel_cuda(TensorIterator& iter) {
   });
 }
 
+template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
+void nanmean_kernel_impl(TensorIterator& iter) {
+  float factor = float(iter.num_output_elements()) / iter.numel();
```
Comment (krshrimali) [inline aten/src/ATen/native/cuda/ReduceMomentKernel.cu:67]: TODO: What is the use of `factor` here?

Also:
`gpu_reduce_kernel<scalar_t, out_t>(iter, MeanOps<acc_t, float> {factor});` - where is `MeanOps` defined?

**Comment by krshrimali** [comment]
**Question**: Is there a way to do this in C++? `~torch.isnan(torch.tensor([1, 2, 3]))`

File: aten/src/ATen/native/ReduceOps.cpp [hunk 4]
```diff
@@ -499,32 +499,29 @@ Tensor& mean_out(Tensor& result, const Tensor& self, DimnameList dim,
...
       at::isFloatingType(scalarType) || at::isComplexType(scalarType),
       "Can only calculate the mean of floating types. Got ",
       toString(scalarType),
       " instead.");
   ScalarType dtype = get_dtype(result, self, opt_dtype, true);
-  Tensor without_nans = torch.where(torch.isnan(self), torch.zeros_like(self), self);
+  Tensor without_nans = at::where(at::isnan(self), at::zeros_like(self), self);
   if (self.device().is_cpu()) {
-    int64_t dim_prod = at::sum(~self.isnan(), axis=dim, keepdims=keepdim, dtype=dtype);
+    at::Tensor dim_prod = at::sum(self.isnan().logical_not(), dim, keepdim, dtype);
```
Comment (krshrimali) [inline aten/src/ATen/native/ReduceOps.cpp:511]: Using `logical_not` as a replacement of `~` operator.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 5]
```diff
@@ -120,6 +122,20 @@ static void mean_kernel_impl(TensorIterator& iter) {
   });
 }
 
+static void nanmean_kernel_impl(TensorIterator& iter) {
+  AT_DISPATCH_ALL_TYPES_AND_C10_COMPLEX(iter.dtype(), "nanmean_cpu", [&] {
+    scalar_t factor = scalar_t(iter.num_output_elements()) / scalar_t(iter.numel());
+    std::cout << "Factor: " << factor << std::endl;
+    std::cout << "Num output elements: " << iter.num_output_elements() << std::endl;
+    std::cout << "Numel: " << iter.numel() << std::endl;
```
Comment (krshrimali) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:130]: Just for testing, TODO: remove this.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 6]
```diff
@@ -1,6 +1,8 @@
 #include <numeric>
 #include <iterator>
 #include <algorithm>
+// TODO: Remove this
+#include <iostream>
```
Comment (krshrimali) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:5]: Included for `std::cout`. Will be removed after testing.

**Comment by krshrimali** [comment]
Initial testing: successful. 🎉 ❤️ Made a PR in the upstream, need to test on CUDA (not possible on my local system :/).
